rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+*   use `strategy.local_results` to get concrete values of results returned by `strategy.schedule`. This may be a blocking call if the result is not yet ready. With any failure that cannot be handled will be ignored and as a result some of the results may be `None`.
+*   call `strategy.join` to wait until all scheduled functions are executed.
+
+```python
+# Connect to remote servers with a user-provided `ClusterResolver` object.
+strategy = ParameterServerStrategyV2(cluster_resolver)
+
+dataset_fn = # a function that returns a dataset
+
+# Clone the dataset on all workers, shuffled with different seeds.
+distributed_dataset = strategy.experimental_distribute_datasets_from_function(
+      dataset_fn)
+
+with strategy.scope():
+  # Create variables on parameter servers in a round-robin fashion.
+  model = create_model()
+  optimizer = tf.keras.optimizers.Adam()
+  accuracy = tf.keras.metrics.CategoricalAccuracy(name="train_accuracy")
+
+  @tf.function
+  def replica_fn(iterator):
+    x, y = next(iterator)
+    with tf.GradientTape() as tape:
+      predictions = model(x, table, training=True)
+      loss = compute_loss(y, predictions)
+    gradients = tape.gradient(loss, model.trainable_variables)
+    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+    accuracy.update_state(y, predictions)
+    return loss
+
+  for _ in range(num_epoches):
+    distributed_iter = iter(distributed_dataset)
+    for i in range(steps_per_epoch):
+      # strategy.schedule pushes a closure in the scheduling queue and
+      # returns a list of future objects immediately.
+      loss = strategy.schedule(replica_fn,
+                               args=(distributed_iter,))
+    strategy.join()
+    model.save()  # save checkpoint/summary...
+    print ("Loss = %f, accuracy = %f" % (
+        strategy.local_results(loss), accuracy.result()))
+```
+
+
+##### Alternative training loop: fully async
+
+Another option from calling `join` every epoch, users can choose to schedule all steps and then asynchronously print metric values. This option doesn’t require any synchronization in epoch boundaries.
+
+
+```python
+# … omitted
+with strategy.scope():
+  # … omitted
+  for _ in range(total_steps)):
+    strategy.schedule(step_fn, args=(iterators,))
+
+  # Print accuracy value every one minute.
+  while not strategy.done():
+    print("Current accuracy: %f" % accuracy.result())
+    time.sleep(60)
+# … omitted
+```
+
+#### Error Reporting From `replica_fn`
+
+Because of the non-blocking `schedule`, any exception raised in `replica_fn` wouldn’t be returned to users immediately. Actually an exception may pollute arbitrary number of functions in flight following the culprit function. We will set the error in returned `Future` objects for the culprit function and these polluted functions and we will raise exceptions when `join` is called.
+
+Therefore the best practice for users is to avoid writing any code that may raise in `replica_fn`:
```

Comment:
This seems like a strange requirement. Do you mean something like, "Try to catch and handle as many raised errors within replica_fn as possible to avoid nullifying the entire scheduled workload"? Is there a way to say, "Something went wrong, and we need to rerun this replica_fn, but not all scheduled jobs"?

Reply:
I think what this meant is more like "If you raise an error within `replica_fn`, you won't be able to handle it by try-except `schedule()`, so the best thing to do is not to raise an error". I understand this is somewhat limiting, and I think this can be solved by having a mechanism of transporting error message from worker threads to client's main thread.

Reply:
In addition to the pitfall of eager async execution that Rick mentioned, there is another gotcha in our eager async implementation: a failing function will pollute all other function scheduled on that executor. For some errors such as NaN errors, we shouldn't do that. We will improve that together with runtime folks. 

cc @haoyuz